openid-4-verifiable-presentations-1_0.md

```diff
@@ -549,11 +555,276 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+future extensions MAY define additional properties that modify the behavior of
+the Wallet when processing the query.
+
+## Credential Query {#credential_query}
+
+A Credential Query is an object representing a request for one specific
+Credential. 
+
+It contains the following properties:
+
+`id`:
+: REQUIRED. A string identifying the Credential in the response and, if provided,
+the constraints in `valid_credential_sets`. The value MUST be a non-empty string
+consisting of alphanumeric, underscore (`_`) or hyphen (`-`) characters.
+Within the Authorization Request, the same `id` MUST NOT
+be present more than once.
+
+`format`:
+: REQUIRED. A string that specifies the format of the requested
+Verifiable Credential. Valid Credential Format Identifier values are defined in
+Appendix A of [@!OpenID.VCI]. The value of this property MUST be one of the
+supported Credential Formats as defined in the Wallet's metadata.
+
+`expect_meta`: 
+: OPTIONAL. An object defining additional properties requested by the Verifier that
+apply to the metadata and validity data of the Credential. The properties of
+this object are defined per Credential Format in (#format_specific_properties). If omitted,
+no specific constraints are placed on the metadata or validity of the requested
+Credential.
+
+`expect_claims`:
+: OPTIONAL. A non-empty array of objects as defined in (#claims_query) that specifies
+claims in the requested Credentials.
+
+`valid_claim_sets`:
+: OPTIONAL. A non-empty array containing arrays of identifiers for elements in
+`expect_claims`. The identifier MAY be postfixed by `?`, indicating that
+delivery of the the respective claim is OPTIONAL.  
+
+## Claims Query {#claims_query}
+
+Each entry in `expect_claims` MUST be an object with the following properties:
+
+`id`:
+: REQUIRED if `valid_claim_sets` is present; OPTIONAL otherwise. A string
+identifying the particular claim. The value MUST be a non-empty string
+consisting of alphanumeric, underscore (`_`) or hyphen (`-`) characters.
+Within the particular `expected_claims` array, the same `id` MUST NOT
+be present more than once.
+
+`path`:
+: REQUIRED if the Credential Format uses a JSON-based claims structure; MUST NOT
+be present otherwise. The value MUST be a claims path pointer that specifies the path to the claim
+within the Verifiable Credential, as defined in (#claims_path_pointer).
+
+`namespace`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the namespace of the claim
+within the Verifiable Credential, e.g., `org.iso.18013.5.1`.
+
+`claim_name`:
+: REQUIRED if the Credential Format is based on mdoc format defined in ISO 18013-5; MUST NOT be present otherwise. 
+The value MUST be a string that specifies the name of the claim within the provided namespace
+in the Verifiable Credential, e.g., `first_name`.
+
+`value`:
```

Comment:
We had a short discussion on this in the previous proposal, but we have not yet addressed this issues that we mentioned there. I'll copy them here:
- Matching on values can be complicated for wallet / credential selection. Because you may not have access to the value of the element at the selection time.
- Even after wallet selection, having access to the values of the data elements may still only be available after user consent because of security reasons (decryption etc). Meaning that the presentation may fail only after user consent.
- Values of data elements are not just strings, integers or booleans. Depending on the credential format they can be maps, arrays, byte strings or other items.
So besides requiring a stricter definition, mostly it's a mechanism that may just not be available because of security or ux reasons.
Since I don't know how to properly address issue 1 and 2 and a solution to issue 3 would be complicated I would not be in favor of having this mechanism since it's not a mechanism RPs can rely on.

Reply:
Any comments on these concerns? Given these issues with data value matching removing the requirements for data value matching is our preferred option. Having a mechanism that is unreliable for the RP to use or could force wallets / credentials selectors to make implementations that compromise security or privacy should be considered very carefully.

Reply:
this has been in the requirements: https://github.com/openid/OpenID4VP/issues/157

Reply:
I don't have any more to add than I already said here:

https://github.com/openid/OpenID4VP/pull/260#discussion_r1759428617

We need to focus on solutions, not on trying to remove functionality that implementors are telling us is needed and that is already present in VP via PE. I've not seen an alternative solutions yet.

Reply:
While generic value matching is a powerful feature, it also runs into the issues that were mentioned.
As mentioned in the call we could make support for the feature optional, which given the potential security and privacy implications is probably a safe bet, but it also means that Relying Parties either cannot rely on the feature or they do and we have bad UX flows.

If we know what specific use cases require data value matching we can come up with a solution that addresses the use case in a way that is reliable from an RP perspective.
An example would be to have Issuer matching, by allowed the RP to select credentials only issued by a particular issuer based on the signing certificate.

Reply:
I think this feature is especially needed in cases where a RP wants to incrementally obtain credential data. Let's assume the RP requests a diploma first. This diploma has the name of its holder (claims based holder binding). The RP will then need to check whether it is actually interacting with this holder. I would expect it to send a request for an identity credential (e.g. a PID or a mDL), scoping the request to the holder's name using the value matching feature.

Reply:
But requiring full access to the credential to do the matching can also have privacy and security implications. Which is why we need to carefully need to balance the pro's and cons of the solutions and see if there are more specific solutions possible for specific use cases.
Taking the example of names also has some issues with value matching. If my credential has my name as "Tom" and the values array is ["tom"] I would not be allowed to return the credential.